### ISSUE TITLE
Unable to reproduce on Davis2017 validation

### ISSUE THREAD

**Issue by junweima** [microsoft/transductive-vos.pytorch#13]
Hi,

I've trained and did inference using the given commands in this repo with 4 GPUs. I trained on DAVIS17 training set and evaluated using the DAVIS17 validation set. I used this repo https://github.com/davisvideochallenge/davis2017-evaluation for evaluation.

but I was only able to get the following result:

```
--------------------------- Global results for val ---------------------------                                                                                                                                     
 J&F-Mean    J-Mean  J-Recall   J-Decay    F-Mean  F-Recall   F-Decay                                                                                                                                              
  0.66228  0.642125  0.758698  0.198392  0.682435  0.807384  0.242924   
```
I trained for 240 epochs as mentioned and evaluated on the 240th epoch. I also evaluated a few checkpoints around 240th but the result was very similar. 

Could you please help me point out what could be the issues?

**Comment by zhirongw** [comment]
It could possibly be the issue of SyncBN implementations. 
Please try the implementation is the following repo: https://github.com/vacancy/Synchronized-BatchNorm-PyTorch.

**Comment by JudasDie** [comment]
> Hi,
> 
> I've trained and did inference using the given commands in this repo with 4 GPUs. I trained on DAVIS17 training set and evaluated using the DAVIS17 validation set. I used this repo https://github.com/davisvideochallenge/davis2017-evaluation for evaluation.
> 
> but I was only able to get the following result:
> 
> ```
> --------------------------- Global results for val ---------------------------                                                                                                                                     
>  J&F-Mean    J-Mean  J-Recall   J-Decay    F-Mean  F-Recall   F-Decay                                                                                                                                              
>   0.66228  0.642125  0.758698  0.198392  0.682435  0.807384  0.242924   
> ```
> 
> I trained for 240 epochs as mentioned and evaluated on the 240th epoch. I also evaluated a few checkpoints around 240th but the result was very similar.
> 
> Could you please help me point out what could be the issues?

Hi, @junweima. I reproduce their code recently. Here is my results after epoch testing,
![image](https://user-images.githubusercontent.com/24988988/87926857-2c17f100-ca37-11ea-83f9-2c962ff49b60.png)

I obtain better results than performance presented in their paper. Moreover, the best result appears at epoch220 in my training.
I would suggest you try epoch testing with the scripts I provided. Please install `mpi4py` in advance.

```
pip install mpi4py
```

Hi, @zhirongw , I creat a pull request about the epoch testing scripts. Would you kindly merge it after checking?

**Comment by zhirongw** [comment]
> Hi,
> 
> I've trained and did inference using the given commands in this repo with 4 GPUs. I trained on DAVIS17 training set and evaluated using the DAVIS17 validation set. I used this repo https://github.com/davisvideochallenge/davis2017-evaluation for evaluation.
> 
> but I was only able to get the following result:
> 
> ```
> --------------------------- Global results for val ---------------------------                                                                                                                                     
>  J&F-Mean    J-Mean  J-Recall   J-Decay    F-Mean  F-Recall   F-Decay                                                                                                                                              
>   0.66228  0.642125  0.758698  0.198392  0.682435  0.807384  0.242924   
> ```
> 
> I trained for 240 epochs as mentioned and evaluated on the 240th epoch. I also evaluated a few checkpoints around 240th but the result was very similar.
> 
> Could you please help me point out what could be the issues?

Hi Junwei,

It turns out that switching to the third party synbn fixes the issue.
We have merged the code to the master branch.
Please retry the newest version.
Thanks.

**Comment by tonysy** [comment]
@junweima @zhirongw  Hi, could you share the environment information of this codebase.
I use the latest commit to train and evaluate DAVIS2017 but got the poor performance, similar to @junweima  initially reported.
![WX20200804-111119](https://user-images.githubusercontent.com/7881589/89248821-3e06b000-d643-11ea-87a8-d7f7aec8e555.png)

**Comment by eastOffice** [comment]
Hi,

We use PyTorch 1.0+ for our experiments. Feel free to let us know if you still have this issue.

Thanks

**Comment by tonysy** [comment]
@eastOffice  I use the pytorch1.4, cuda10.0.  I have run the latest version of the code twice but got similar poor results. Could you share the specific version of PyTorch and Cuda in detail, thanks  a lot?

**Comment by eastOffice** [comment]
Hi,

We used PyTorch 1.0 but 1.4 also worked, and cuda 10.0. You can try 1.0 and make sure that you use this Sync BN implementation: https://github.com/vacancy/Synchronized-BatchNorm-PyTorch.

**Comment by tonysy** [comment]
@eastOffice 

Hi, I have tried to reproduce the code based on the pytorch 1.0, where I can get better results compared with pytorch1.4.
But still less than the performance reported in your README. The software version may matter a lot, if you share share more details about your environment, thanks a lot.

```

Per-sequence results saved in ../saves/davis2017_train_baseline_240e_4gpu/epoch2
39/per-sequence_results-val.csv
--------------------------- Global results for val ---------------------------
 J&F-Mean    J-Mean  J-Recall   J-Decay    F-Mean  F-Recall   F-Decay
 0.702935  0.678205  0.797827  0.160502  0.727665  0.865816  0.186022
```
 
> Hi,
> 
> We used PyTorch 1.0 but 1.4 also worked, and cuda 10.0. You can try 1.0 and make sure that you use this Sync BN implementation: https://github.com/vacancy/Synchronized-BatchNorm-PyTorch.

@JudasDie  Hi, would you like to share your environment information for reproducing the performance, thanks in advance.

**Comment by fatemehazimi990** [comment]
Hi, thanks for sharing the code, and also thanks to others for sharing their experience. I also struggled with a similar problem and seems the main issue is with the torch version (I tried the suggested batch norm implementation with multiple torch versions newer than 1.0 and the problem was still there). Could be that the default behavior of PyTorch has changed, but in any case, it would be super helpful to find the main cause and would make it easier to use the model as a baseline. If anyone is interested, I would be happy to collaborate in running experiments :) @eastOffice
